Fix issues with output reloading

ycombinator opened the pull request:
## What does this PR do?

Beats Central Management (today) and Agent (in the future) will need to reload outputs on the fly, whenever users change the Beat's configuration while the Beat is running. The current output reloading implementation has three bugs:

1. When outputs are reloaded several times, and very quickly at that, all events don't always getting eventually published,

1. Whenever an output is reloaded, a goroutine is leaked from the publisher pipeline's output handling code, and

2. If the output being reloaded is the Elasticsearch output, old TCP connections to Elasticsearch are hanging around.

This PR addresses all three issues.

## Why is it important?

- To ensure that all events will eventually be published, even if outputs are rapidly reloaded multiple times.
- To prevent leakage of memory (goroutines consume memory).
- To prevent exhaustion of file descriptors over time.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~I have made corresponding changes to the documentation~
- [ ] ~I have made corresponding change to the default configuration files~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

1. Start up Elasticsearch and Kibana and setup [Beats Central Management](https://www.elastic.co/guide/en/beats/filebeat/current/configuration-central-management.html).
   1. Enroll Filebeat. This will backup your original `filebeat.yml` file and create a new one suitable for Beats Central Management.
   1. Create a configuration tag in the Beats Central Management UI in Kibana with input paths containing `/tmp/logs/*.log` and the Elasticsearch output pointing to localhost:9200 (or wherever your local ES is listening). Make sure to assign this tag to your enrolled Filebeat instance in the UI.
3. Edit the new `filebeat.yml` file and set `management.period: 2s` to speed up output reloading.
4. Start up Filebeat with the `-httpprof localhost:5050` flag. This will let us see if Filebeat is leaking goroutines.
5. In a new window, check TCP connections used by your Filebeat process. This will let us see if Filebeat is leaking file descriptors.
   ```
   watch -n 2 lsof -a -cfilebeat -i4 -i6 -itcp
   ```
6. In another new window, make sure some data is constantly being ingested by Filebeat from the input path.
   ```
   rm -rf /tmp/logs
   mkdir -p /tmp/logs
   while true; do echo "foo bar $RANDOM" >> /tmp/logs/test.log; sleep 1; done
   ```
7. In the Central Management UI, change the configuration for your tag slightly (maybe add an input path then later remove it and so on) to trigger output reloading in Filebeat. 
8. Check the Filebeat log for messages like `New configurations retrieved` to confirm that your changes in Kibana are being picked up by Filebeat.
8. Check the output of the `lsof` window and make sure the list of file descriptors isn't growing over time.
9. Visit (or refresh) http://localhost:5050/debug/pprof/goroutine?debug=1 in a browser window and make sure the number of goroutines isn't growing over time.
10. Wait a few seconds, then repeat steps 6 onwards a few times.

## Related issues

- Fixes elastic/beats#10491

ycombinator commented on libbeat/esleg/eslegclient/connection.go:
```diff
@@ -407,6 +407,9 @@ func (conn *Connection) execHTTPRequest(req *http.Request) (int, []byte, error)
 	}
 
+	// TODO: workaround for output reloading leaking FDs until context.WithCancel is used on transport dialer instead
+	req.Header.Set("Connection", "close")
```
Before I made this workaround/hack, I tried two different approaches to fix the FD leak the "right way":

1. Just as there is an `http.Request` function (that we currently use in the `eslegclient` code), there is also a `http.RequestWithContext` function. I replaced our usage with the latter one, created a `context.WithCancel` inside `eslegclient.NewConnection` and used the cancel function returned to cancel the context from the `eslegclient.(*Connection).Close()` method. This is essentially the same fix that @urso tried in #10599. However, when I used `lsof` to monitor FDs being used by my Filebeat process, I noticed that this change had no effect!

2. So then, thinking about it some more, I realized the problem wasn't so much on canceling the HTTP requests but it was more to do with closing the underlying TCP connection. To fix this the "right way", I tried threading `context.Context` through the various `Dialer`s (which would allow us to call `DialContext` instead of `Dial` in the right places). This turned out to be quite a complicated and large refactoring! So instead, I settled for this workaround here. I think we do need to undertake this refactoring at some point but I'm not sure it's worth blocking progress at the moment?

urso replied:
This would force the TCP connection to be dropped after each bulk request. Internally the go HTTP package uses a connection pool (actually the `Transport` type does) and keeps the HTTP session alive in order to send more requests. Would be nice if we could keep this.

This makes me wonder if this is really a leak, or just an open connection that eventually gets closed in a few minutes by the go runtime. The [`Transport`](https://pkg.go.dev/net/http?tab=doc#Transport) type has a few settings for controlling the connection pool. For example `IdleConnTimeout` (which we must configure). On shutdown one can also call `CloseIdleConnections`.
I think we should always configure a timeout (and expose it as a setting) + call `CloseIdleConnections` (best effort) approach to close connections early.

We also have some helpers for working with contexts in this package: https://pkg.go.dev/github.com/elastic/go-concert/ctxtool. This allows us to hook up contexts with custom functions, plain go channels, or merge cancelation from multiple contexts.

ycombinator replied:
Thanks @urso. 

> This makes me wonder if this is really a leak, or just an open connection that eventually gets closed in a few minutes by the go runtime.

Yes, good point. I will run my tests for longer and see if the connection FDs eventually get released.

> I think we should always configure a timeout (and expose it as a setting) + call CloseIdleConnections (best effort) approach to close connections early.

++ to this approach, will implement.

> We also have some helpers for working with contexts in this package: https://pkg.go.dev/github.com/elastic/go-concert/ctxtool. This allows us to hook up contexts with custom functions, plain go channels, or merge cancelation from multiple contexts.

I will look into this as well, but probably part of a follow up PR.

ph replied:
++ Will all the above that @urso mentioned :)

ycombinator replied:
Implemented a new setting in `eslegclient.ConnectionSettings` named `IdleConnTimeout`. The name matches the setting on`http.Transport`. Default value is 1 minute.

Also implemented calling `CloseIdleConnections()` from `eslegclient.(*Connection).Close()`.

Net result is that we don't see FDs sticking around in the test.

elasticmachine commented:
Pinging @elastic/integrations-services (Team:Services)

urso commented on libbeat/publisher/pipeline/output.go:
```diff
@@ -18,115 +18,130 @@
+		select {
 
+		case <-w.done:
+			return
```
Not sure, just a hunch. But please check the state of batches and the workqueue in case we return early. Is the workqueue reused or do we create a new workqueue for the new set of outputs? In the former case we should be fine. But the original code did wait for the work queue to be closed. This way we've been sure to not have any batches in flight. By returning early there is a chance of having batches still in the workqueue. Is the queue drained somewhere else? At some point (maybe a defer here) we need a draining loop like this:

```
// cancel -> return to retryer -> rescheduled batches with new queue
for batch := range w.qu {
  batch.Cancelled()
}
```

or like this:

```
// transfer pending batches to new workqueue
for batch := range w.qu {
  newQueue <- batch
}
```

ycombinator replied:
Okay, I checked this and it looks like a **new workqueue** is created each time we set a new group of outputs on the controller:

https://github.com/elastic/beats/blob/95626b8f1690344312c0831ab2bdcbccffe4d089/libbeat/publisher/pipeline/controller.go#L103

The retryer is then given this new workqueue (assigned to `retryer.out`). So canceling the batch from the old worker so it's returned to the retry queue (`retryer.in`) makes sense here.

ycombinator replied:
Implemented in 400a5d4d4.

urso commented:
Would be nice if we could introduce an unit test trying to replace an output and check all batches are eventually handled.

ycombinator commented:
> Would be nice if we could introduce an unit test trying to replace an output and check all batches are eventually handled.

I'm actually going to implement unit tests on the original code first to ensure they keep passing with the current PR. I've done this in https://github.com/elastic/beats/pull/17460. Please take a look when you get a chance. 

Once  https://github.com/elastic/beats/pull/1746 is reviewed and merged, I'll rebase this PR on it.

ycombinator commented:
Unit tests (per the previous two comments) have now been added in #17460. I will now resume working on this PR, making sure the same tests continue to pass with the changes in this PR and also address review feedback.

ycombinator commented:
I've addressed @urso's review feedback and also added another unit test in this PR specifically to test output reloading. That test is currently failing sometimes so I need to investigate why and fix.

ycombinator commented:
> ... added another unit test in this PR specifically to test output reloading. That test is currently failing sometimes so I need to investigate why and fix.

While investigating this I found that this new unit test isn't passing on `master` either (i.e. with the original implementation). If the output is reloaded multiple times very quickly, not all events get published. To figure out what's going on on `master` and resolve it first, I've created a new PR just with this new unit test: https://github.com/elastic/beats/pull/17657.

ph commented:
@ycombinator Any idea how to fix it?

ycombinator commented:
> Any idea how to fix it?
@ph Not 100% but I made some progress in #17657. After going over it with @urso on Tuesday (I was off yesterday), my next steps here are to bring the unit test written in #17567 into this PR and then try to make it pass using the code fixes in this PR, maybe some that were experimented with in #17567, and maybe some others still that we haven't discovered yet. 

The general problem seems to be that there's a race condition somewhere in the publisher code that's being exposed when the output controller's `Set` method is called quickly and repeatedly (which could be the case with Agent). This race condition prevents all events from eventually being published, but only some times.

ycombinator commented on libbeat/publisher/pipeline/consumer.go:
```diff
@@ -154,7 +154,7 @@ func (c *eventConsumer) loop(consumer queue.Consumer) {
 
 		paused = c.paused()
-		if !paused && c.out != nil && batch != nil {
+		if c.out != nil && batch != nil {
```
This turned out to be a key change in getting the test to pass.

Consider the case where the consumer is paused (so `paused == true` here) but it still has a batch of events that it has consumed from the queue but not yet sent to the workqueue. In this scenario because of the `!paused` clause here, we would fall into the `else` clause, which would set `out = nil`. This would then cause the last `select` in this method to block sometimes (depending on timing). 

Now, if we have a valid output group `c.out` and a `batch`, we don't pay attention to the `paused` state; this ensures that this batch will be send to that output group's work queue via that final `select`'s final `case`.

ph replied:
Oh wow nice fix!

ycombinator commented on libbeat/publisher/pipeline/controller.go:
```diff
@@ -34,7 +34,8 @@ type outputController struct {
 
-	queue queue.Queue
+	queue     queue.Queue
+	workQueue workQueue
```
This is another change I made, to simplify things. Now the output controller creates a work queue in it's constructor and the same one is used by all output workers across time.

ph replied:
good change!

ycombinator commented:
@urso @ph I finally managed to figure out the root cause of the race condition that was making the new unit test in this PR sometimes fail. It had to do with a timing issue where the consumer was getting paused after a batch was consumed from the "main" queue but before it was sent to the work queue. 

Please take a look at the changes I've made and let me know what you think. Thanks!

ycombinator commented:
Hmmm, looks like something's not right with the fix I made 😞. Even with just 1 `testing/quick` iteration, the test times out in Travis. I was able to run it locally (and get it to pass) with 250 iterations! 🤔

ycombinator commented:
@urso @ph This PR is ready for another round of review. The CI failures with `libbeat` are unrelated (see https://github.com/elastic/beats/pull/17819). Thanks!

urso commented on libbeat/publisher/pipeline/controller.go:
```diff
@@ -65,15 +67,16 @@ func newOutputController(
+		beat:      beat,
+		monitors:  monitors,
+		observer:  observer,
+		queue:     b,
```
that's funny. Why is the parameter named `b` ? As we've modified this piece of code, leats clean up the naming a little :)

ycombinator replied:
Fixed in 5ff7219.

urso commented on libbeat/publisher/pipeline/controller.go:
```diff
@@ -136,12 +137,13 @@ func (c *outputController) Set(outGrp outputs.Group) {
 
 	// restart consumer (potentially blocked by retryer)
 	c.consumer.sigContinue()
+	c.consumer.sigUnWait()
```
`sigUnWait` is used by the retryer, no? Why add it here?

ycombinator replied:
I found that this was needed to make the `TestReload` test pass. Without it the retryer never un-waited the consumer by itself and so events stopped being consumed and sent to the output.

That said, I agree that this is something the retryer should automatically call internally rather than "leak" this responsibility to the controller. I'll look into why it's not doing that.

urso replied:
Yeah, this sounds like a potential bug to me.

ycombinator replied:
Addressed in 6a6680c.

urso commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -51,20 +44,27 @@ func TestMakeClientWorker(t *testing.T) {
+				defer retryer.close()
 
 				var published atomic.Uint
+				var batchCount int
```
I used `batchCount` for debugging. We can remove it I think.

ycombinator replied:
Removed in 15a1383.

urso commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -85,41 +85,51 @@ func TestMakeClientWorker(t *testing.T) {
 
 			err := quick.Check(func(i uint) bool {
-				numBatches := 1000 + (i % 100) // between 1000 and 1099
+				numBatches := 10000 + (i % 100) // between 1000 and 1099
```
oops, I guess that was me. Want to to reduce it to `1000` again to reduce the running time?

ycombinator replied:
Reduced in 6e10331.

urso commented on libbeat/publisher/pipeline/output_test.go:
```diff
@@ -85,41 +85,51 @@ func TestMakeClientWorker(t *testing.T) {
 
 				var publishedFirst atomic.Uint
 				blockCtrl := make(chan struct{})
+				var batchCount int
```
might be a debugging variable that can be removed.

ycombinator replied:
Removed in 15a1383.

ycombinator commented:
I've addressed the minor "cleanup" points from the last round of review. The major point that remains to be addressed is https://github.com/elastic/beats/pull/17381#discussion_r411422026. I will do that tomorrow.

ycombinator commented:
@urso I've addressed all your latest feedback. Please re-review when you get a chance. Thanks!

ycombinator commented:
CI failures are unrelated. Libbeat builds in both Jenkins and Travis CI are green. Merging.
